design-documents/declarative-schema-database-compatibility.md

```diff
@@ -0,0 +1,324 @@
+# Declarative schema database compatibility
+
+## About
+
+Declarative schema relies on querying data that may have different format depending on database. For example, here is the query from `Magento\Framework\Setup\Declaration\Schema\Db\MySQL\DbSchemaReader::readColumns`
+
+```
+SELECT 
+    COLUMN_NAME as 'name',
+    COLUMN_DEFAULT as 'default',
+    DATA_TYPE as 'type',
+    IS_NULLABLE as 'nullable',
+    COLUMN_TYPE as 'definition',
+    EXTRA as 'extra',
+    COLUMN_COMMENT as 'comment'
+FROM
+    information_schema.COLUMNS
+WHERE
+    TABLE_SCHEMA = 'magento'
+AND
+    TABLE_NAME = 'store_website'
+ORDER BY
+    ORDINAL_POSITION ASC;
+```
+
+MySQL returns NULL for COLUMN_DEFAULT and MariaDB returns NULL as a string.
+
+## Design
+
+Introduce new interfaces under Db namespace that allow to manage table schema, refactor implementations of the following interfaces in declarative schema to use interfaces under Db namespace.
+
+Similar approach need to be taken for clients of the following interfaces
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DbSchemaReaderInterface`
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DbSchemaWriterInterface`
+
+Revise implementation of these interfaces to see if anything need to be changed here
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DDLTriggerInterface`
+1. `\Magento\Framework\Setup\Declaration\Schema\Db\DbDefinitionProcessorInterface` (see usages of `\Magento\Framework\Setup\Declaration\Schema\Db\DefinitionAggregator`)
+
+New interfaces and types for managing table schema.
+
+```
+class ConstraintType extends \SplEnum
+{
+    const primary = 'primary';
+    const unique = 'unique';
+    const auto_increment = 'auto_increment';
+}
+```
+
+```
+class IndexType extends \SplEnum
+{
+    // For MySQL, Oracle, MS SQL and Postgres will use btree
+    const normal = 'normal';
+
+    // For MySQL and MS SQL fulltext, one of text indexes for Oracle and text for Postgres
+    const text = 'text';
+}
+```
+
+```
+class OnDelete extends \SplEnum
```

Comment:
Won't these enums be considered APIs int the future? It will be hard to add new values after that. Perhaps we should just have arrays of strings set with DI configuration instead of static enums

Reply:
Enums allow to restrict set of values, it helps to make sure that all values supported by all databases.